File: src/java_answer/Lessons12909.java
package java_answer;

/**
 * @author dave.th
 * https://programmers.co.kr/learn/courses/30/lessons/12909
 */
public class Lessons12909 {
    public static void main(String[] args) {
        System.out.println((new Lessons12909()).solution("()()"));
    }

    private boolean solution(String s) { // 스택 사용시 시간초과떠서 카운팅 변수 하나만 사용
        int count = 0;
        char[] arr = s.toCharArray();
        for(int i=0; i<arr.length; i++) {
            if("(".charAt(0) == arr[i]) {
                count++;
            } else {
                if(count == 0) return false;
                count--;
            }
            if(count > arr.length - i) return false;
        }
        return count == 0;
    }
}

File: src/scala_answer/Lessons12909.scala
package scala_answer

import scala.collection.mutable

/**
  * @author dave.th
  * https://programmers.co.kr/learn/courses/30/lessons/12909
  */
object Lessons12909 {
  def main(args: Array[String]) : Unit = {
    println(solution("()()"))
  }

  def solution(s : String) : Boolean = {
    val stack = new mutable.Stack[Char]

    for(i <- s.toCharArray) {
      if("(".charAt(0).equals(i)) {
        stack.push(i)
      } else {
        if(stack.isEmpty) return false
        stack.pop
      }
    }

    stack.isEmpty
  }
}
